libp2p/switch.nim

```diff
@@ -486,19 +487,22 @@ proc start*(s: Switch): Future[seq[Future[void]]] {.async, gcsafe.} =
     except CatchableError as exc:
       trace "Exception occurred in Switch.start", exc = exc.msg
 
-  var startFuts: seq[Future[void]]
   for t in s.transports: # for each transport
     for i, a in s.peerInfo.addrs:
       if t.handles(a): # check if it handles the multiaddr
-        var server = await t.listen(a, handle)
+        # We spawning listeners and store all of them inside `switch`.
+        let listener = t.listen(a, handle)
         s.peerInfo.addrs[i] = t.ma # update peer's address
-        startFuts.add(server)
+        s.listeners.add(listener)
 
   if s.pubSub.isSome:
     await s.pubSub.get().start()
 
   info "started libp2p node", peer = $s.peerInfo, addrs = s.peerInfo.addrs
-  result = startFuts # listen for incoming connections
+  
+proc join*(s: Switch): Future[void] =
+  ## Waits for switch `s` to be finished.
+  allFutures(s.listeners)
```

Comment:
If one of the transports failed to start or failed after some time it will complete/fail `Future[void]` stored inside of `switch.listeners`, and because of that `allFutures` will not wait on completion of this failed transport.

The previous mechanism returns sequence of anonymous `Future[void]` which require you to always use `allFutures/allFinished` to wait for transports completion/failure/finish. But it was also impossible to wait for particular transport with specified multiaddress.

Reply:
> The previous mechanism returns sequence of anonymous Future[void] which require you to always use allFutures/allFinished to wait for transports completion/failure/finish. But it was also impossible to wait for particular transport with specified multiaddress.

This is a good case for `all` or `any` for example. If one failed/comletted, the caller can decide what to do next, to ignore or fail. Not a perfect mechanism still, but better than hiding it away.

> So in previous mechanism there also was no way to recover server instance from Future[T] returned by "old" transport.listen() call.

Yes and we probably need to allow it somehow, the problem with the `join` however, is that it hides it and there is no way to know what happened. IMO, we should just leave it as is for now and continue returning a seq of futures from `switch.start`, and work on improving the mechanism in a separate PR.

Reply:
@dryajov there no breaking changes in this PR. The only thing which is present - cosmetic changes in API, which make interface more understandable and easy to use.

For example if you want to keep "old" method you can easily do it with this PR:

```nim
await allFuturesThrowing(switch.listeners)
```
But you also can use "new" method using `join()` procedure.

One more reason to change is that now this list of listeners is stored inside of switch and you do not need to create closures or temporary variables or pass this list as an argument, everything is now stored inside of `switch` object.

For example (this is comparison of old test and new test):

```nim
  var nodes = generateNodes(2)
  var awaiters: seq[Future[void]]
  awaiters.add((await nodes[0].start()))
  awaiters.add((await nodes[1].start()))
  #
  # some test's logic here
  #
  await allFuturesThrowing(nodes[0].stop(), nodes[1].stop())
  await allFuturesThrowing(awaiters)
```

And new one

```nim
  var nodes = generateNodes(2)
  await allFuturesThrowing(nodes[0].start(), nodes[1].start())
  #
  # some tests's logic here
  #
  await allFuturesThrowing(nodes[0].stop(), nodes[1].stop())
  await allFuturesThrowing(nodes[0].join(), nodes[1].join())
```

As you can see in this examples, there no longer any need to use temporary variable `awaiters`. But if you still need to use "old" logic you can use this example:

```nim
  var nodes = generateNodes(2)
  await allFuturesThrowing(nodes[0].start(), nodes[1].start())
  #
  # some tests's logic here
  #
  await allFuturesThrowing(nodes[0].stop(), nodes[1].stop())
  await allFuturesThrowing(nodes[0].listeners & nodes[1].listeners)
```

All 3 examples perform same task and able to achieve same result, the only difference that after merging this PR you will be able to use one more approach. So API become more ergonomic.

*P.S.* If you do not like name `join()` we can rename it easily.